File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/client/HttpClient.java
package com.ficture7.aasexplorer.client;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.net.URI;

import static com.ficture7.aasexplorer.util.ObjectUtil.checkNotNull;

/**
 * Represents a {@link Client} which uses HTTP as transport.
 *
 * @author FICTURE7
 */
public abstract class HttpClient implements Client {

    /**
     * User-Agent used in the headers of requests made by {@link #get(URI)}.
     */
    protected static final String USER_AGENT = "aasexplorer-core-" + HttpClient.class.getPackage().getImplementationVersion();

    /**
     * Timeout used for the requests made by {@link #get(URI)},
     */
    protected static final int TIMEOUT = 5000;

    /**
     * Returns the HTML page at the specified {@link URI} in the form of a {@link Document}.
     *
     * @param uri {@link URI} of the HTML page.
     * @return HTML page at the specified {@link URI} in the form of a {@link Document}.
     * @throws NullPointerException {@code uri} is null.
     * @throws DownloadException    Exception caught when downloading the HTML page.
     */
    protected Document get(URI uri) throws DownloadException {
        checkNotNull(uri, "uri");

        String uriString = uri.toString();
        try {
            return Jsoup.connect(uriString)
                        .timeout(TIMEOUT)
                        .userAgent(USER_AGENT)
                        .get();
        } catch (Exception e) {
            throw new DownloadException(uriString, e);
        }
    }

    /**
     * Parses the specified {@link Document} instance and uses the specified {@link Processor}
     * to process parsed data from the {@link Document} instance.
     *
     * @param document {@link Document} instance.
     * @param processor {@link Processor} instance.
     * @throws ParseException Exception caught when parsing the {@link Document}.
     */
    protected abstract void parse(Document document, Processor processor) throws ParseException;

    /**
     * Represents a {@link Document} processor.
     *
     * @author FICTURE7
     */
    protected static abstract class Processor {

        /**
         * Processes the specified name and URL.
         * @param name Name.
         * @param url URL.
         */
        public abstract void process(String name, String url);
    }
}

File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/client/HttpClient.kt
package com.ficture7.aasexplorer.client

import org.jsoup.Jsoup
import org.jsoup.nodes.Document

import java.net.URI

/**
 * Represents a [Client] which uses HTTP as transport.
 *
 * @author FICTURE7
 */
abstract class HttpClient : Client {

    /**
     * Returns the HTML page at the specified [URI] in the form of a [Document].
     *
     * @param uri [URI] of the HTML page.
     * @return HTML page at the specified [URI] in the form of a [Document].
     * @throws NullPointerException `getURI` is null.
     * @throws DownloadException    Exception caught when downloading the HTML page.
     */
    @Throws(DownloadException::class)
    protected operator fun get(uri: URI): Document {
        val uriString = uri.toString()
        try {
            return Jsoup.connect(uriString)
                    .timeout(TIMEOUT)
                    .userAgent(USER_AGENT)
                    .get()
        } catch (e: Exception) {
            throw DownloadException(uriString, e)
        }
    }

    /**
     * Parses the specified [Document] instance and uses the specified [Processor]
     * to process parsed data from the [Document] instance.
     *
     * @param document [Document] instance.
     * @param processor [Processor] instance.
     * @throws ParseException Exception caught when parsing the [Document].
     */
    @Throws(ParseException::class)
    protected abstract fun parse(document: Document, processor: Processor)

    /**
     * Represents a [Document] processor.
     *
     * @author FICTURE7
     */
    protected abstract class Processor {
        /**
         * Processes the specified getName and URL.
         * @param name Name.
         * @param url URL.
         */
        abstract fun process(name: String, url: String)
    }

    companion object {
        /**
         * User-Agent used in the headers of requests made by [.get].
         */
        protected val USER_AGENT = "aasexplorer-core-" + HttpClient::class.java.`package`.implementationVersion

        /**
         * Timeout used for the requests made by [.get],
         */
        protected val TIMEOUT = 5000
    }
}
